File: android/src/main/kotlin/com/freemovevr/google_mediapipe_pose_detection/PoseDetector.kt
package com.freemovevr.google_mediapipe_pose_detection

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import android.util.Log
import com.google.mediapipe.framework.image.MPImage
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.core.Delegate
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarker
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarkerResult
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler


class PoseDetector(
    private val context: Context,
    private val flutterPluginBinding: FlutterPlugin.FlutterPluginBinding
) : MethodCallHandler, EventChannel.StreamHandler {

    // Event channel for sending live stream pose detection results back to Flutter
    private var eventSink: EventChannel.EventSink? = null

    // Mediapipe's Pose Landmark task object
    private var poseLandmarker: PoseLandmarker? = null

    // Internal check for if the poseLandmarker is currently processing an image
    private var isWorking = false

    // The next image for the landmarker to detect
    private var savedImage: MPImage? = null

    /**
     * Method for Flutter to interact with Kotlin by specifying what method it wants to invoke.
     *
     * @param call The contents received from Flutter
     * @param result To be used to return a value
     */
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            START -> handleDetection(call, result)
            CLOSE -> {
                closeDetector()
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    /**
     * Check if the landmarker has been created yet, and if not, create it.
     * If call specifies this detection as a single image, process the image.
     * If it is a live stream, save the image at savedImage for detectSavedImage() to process when
     * the landmarker is available. Override the existing savedImage if one already exists
     *
     *
     * @param call The contents received from Flutter
     * @param result To be used to return a value
     */
    private fun handleDetection(call: MethodCall, result: MethodChannel.Result) {

        // Convert the image now so we do not slow down detectSavedImage() and the detectionThread
        val imageData = call.argument<Any>("imageData") as Map<*, *>
        val inputImage: MPImage = getInputImageFromData(
            imageData,
            result
        ) ?: return

        // If pose detector is null, try to set it up
        val options = call.argument<Map<String, Any>>("options")
        if (options == null) {
            result.error("PoseDetectorError", "Invalid options", null)
            return
        }
        poseLandmarker = poseLandmarker ?: setupPoseLandmarker(options)

        // If it is still null, we cannot detect the poses
        if (poseLandmarker == null) {
            result.error("PoseDetectorError", "Could not make landmarker", null)
            return
        }

        when (options["mode"] as String?) {
            // If we are just processing one image, we can do that here and return the result
            RUNNING_MODE_IMAGE -> {
                isWorking = true
                poseLandmarker?.detect(inputImage)?.also { landmarkResult ->

                    val convertedLandmarkResult: MutableList<List<Map<String, Any>>> = convertPoseData(landmarkResult)

                    isWorking = false

                    result.success(convertedLandmarkResult)
                }
            }

            // If we have livestream data, just save the image for the detectionThread
            RUNNING_MODE_LIVESTREAM -> {

                savedImage = inputImage
            }
        }
        // We must always end by calling result.
        // If we have gotten here we do not need to do anything more,
        // so just return successfully with nothing.
        result.success(null)
    }

    /**
     * If the poseLandmarker is not currently processing an image, detect the current savedImage
     * and set savedImage to null
     */
    fun detectSavedImage()
    {
        if(!isWorking && savedImage != null)
        {
            isWorking = true

            val frameTime = SystemClock.uptimeMillis()
            poseLandmarker!!.detectAsync(savedImage, frameTime)
            savedImage = null
        }
    }

    /**
     * Attempts to set up a poseLandmarker with the settings provided.
     *
     * @param options a map of how to set up the pose landmarker
     */
    private fun setupPoseLandmarker(options: Map<String, Any>): PoseLandmarker? {
        // Set general pose landmarker options
        val baseOptionBuilder = BaseOptions.builder()

        // Use the specified hardware for running the model. Default to CPU
        when (options["delegate"] as String?) {
            DELEGATE_CPU -> {
                baseOptionBuilder.setDelegate(Delegate.CPU)
            }

            DELEGATE_GPU -> {
                baseOptionBuilder.setDelegate(Delegate.GPU)
            }
        }

        val modelName = "assets/" +
                when (options["model"] as String?) {
                    MODEL_POSE_LANDMARKER_FULL -> "pose_landmarker_full.task"
                    MODEL_POSE_LANDMARKER_LITE -> "pose_landmarker_lite.task"
                    MODEL_POSE_LANDMARKER_HEAVY -> "pose_landmarker_heavy.task"
                    else -> "pose_landmarker_full.task"
                }

        val assetPath: String = flutterPluginBinding
            .flutterAssets
            .getAssetFilePathBySubpath(modelName, "google_mediapipe_pose_detection")

        baseOptionBuilder.setModelAssetPath(assetPath)

        try {
            val baseOptions = baseOptionBuilder.build()
            // Create an option builder with base options and specific
            // options only use for Pose Landmarker.
            val optionsBuilder =
                PoseLandmarker.PoseLandmarkerOptions.builder()
                    .setBaseOptions(baseOptions)
                    .setMinPoseDetectionConfidence((options["minPoseDetectionConfidence"] as Double).toFloat())
                    .setMinTrackingConfidence((options["minTrackingConfidence"] as Double).toFloat())
                    .setMinPosePresenceConfidence((options["minPosePresenceConfidence"] as Double).toFloat())

            val runningMode = options["mode"] as String?

            when (runningMode) {
                RUNNING_MODE_IMAGE -> {
                    optionsBuilder.setRunningMode(RunningMode.IMAGE)
                }

                RUNNING_MODE_LIVESTREAM -> {
                    optionsBuilder.setRunningMode(RunningMode.LIVE_STREAM)
                }
            }

            // The ResultListener and ErrorListener only use for LIVE_STREAM mode.
            if (runningMode == RUNNING_MODE_LIVESTREAM) {
                optionsBuilder
                    .setResultListener(this::returnLivestreamResult)
                    .setErrorListener(this::returnLivestreamError)
            }

            val modelOptions = optionsBuilder.build()
            poseLandmarker =
                PoseLandmarker.createFromOptions(context, modelOptions)

            return poseLandmarker

        } catch (e: IllegalStateException) {
            Log.e(
                TAG,
                "Pose Landmarker failed to initialize. See error logs for details"
            )
            Log.e(
                TAG, "MediaPipe failed to load the task with error: " + e.message
            )
        } catch (e: RuntimeException) {
            // This occurs if the model being used does not support GPU
            Log.e(
                TAG,
                "Pose Landmarker failed to initialize. This occurs if the model being used " +
                        "does not support GPU."
            )
            Log.e(
                TAG,
                "Image classifier failed to load model with error: " + e.message
            )
        }
        return null
    }

    /**
     * The callback the poseLandmarker makes after
     * successfully processing an image from a livestream.
     * Required function for Mediapipe, SHOULD NOT be used directly.
     * Once called, the results are sent back to Flutter through the event channel.
     */
    private fun returnLivestreamResult(
        result: PoseLandmarkerResult,
        image: MPImage
    ) {
        isWorking = false

        val landmarkList: MutableList<List<Map<String, Any>>> = convertPoseData(result)

        // Flutter requires the event channel to be called on the UI thread
        Handler(Looper.getMainLooper()).post {
            eventSink?.success(landmarkList)
        }
    }

    /**
     * The callback the poseLandmarker makes after
     * unsuccessfully processing an image from a livestream.
     * Required function for Mediapipe, SHOULD NOT be used directly.
     * If we get an error, just try to log it.
     */
    private fun returnLivestreamError(error: RuntimeException) {
        Log.e(
            TAG,
            error.message ?: "An unknown error has occurred"
        )
    }

    /**
     * Takes in the results from the poseLandmarker,
     * and converts it to be readable for Flutter's supported platform channel data types
     */
    private fun convertPoseData(poseLandmarkerResult: PoseLandmarkerResult): MutableList<List<Map<String, Any>>> {

        val landmarkArray: MutableList<List<Map<String, Any>>> =
            ArrayList()
        if (poseLandmarkerResult.landmarks().isNotEmpty()) {
            for (pose in poseLandmarkerResult.landmarks()) {
                val landmarks: MutableList<Map<String, Any>> =
                    ArrayList()
                for ((index, landmark) in pose.withIndex()) {
                    val landmarkMap: MutableMap<String, Any> =
                        HashMap()
                    landmarkMap["type"] = index
                    landmarkMap["x"] = landmark.x()
                    landmarkMap["y"] = landmark.y()
                    landmarkMap["z"] = landmark.z()
                    landmarkMap["presence"] = landmark.presence().orElse(0.0F)
                    landmarkMap["visibility"] = landmark.visibility().orElse(0.0F)
                    landmarks.add(landmarkMap)
                }
                landmarkArray.add(landmarks)
            }
        }
        return landmarkArray
    }

    /**
     * Cleanup function.
     */
    private fun closeDetector() {
        poseLandmarker?.close()
    }

    companion object {
        // Logging tag
        const val TAG = "google_mediapipe_pose_detection"

        // Calls Flutter can make to Android
        private const val START = "startPoseDetector"
        private const val CLOSE = "closePoseDetector"

        // Configuration options for the pose landmarker
        const val DELEGATE_CPU = "CPU"
        const val DELEGATE_GPU = "GPU"
        const val RUNNING_MODE_IMAGE = "image"
        const val RUNNING_MODE_LIVESTREAM = "liveStream"
        const val MODEL_POSE_LANDMARKER_FULL = "full"
        const val MODEL_POSE_LANDMARKER_LITE = "lite"
        const val MODEL_POSE_LANDMARKER_HEAVY = "heavy"
    }

    /**
     * Required method for StreamHandler.
     * Sets up the eventSink to be used in relaying pose detection results back to Flutter
     */
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    /**
     * Required method for StreamHandler.
     */
    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}

File: ios/Classes/PoseDetector.swift
//
//  PoseDetector.swift
//  google_mediapipe_pose_detection
//
//  Manages reading frames and sending back landmark data using Mediapipe.
//
//  Created by William Parker
//

import Foundation
import Flutter
import MediaPipeTasksVision


/// Main class of the library. Does all the work to convert the frames recived from Flutter into landmark data.
public class PoseDetector: NSObject, FlutterPlugin, FlutterStreamHandler
{
    
    /// A callback to respond with landmark data.
    var eventSink: FlutterEventSink?
    
    /// The Mediapipe Landmarker object to detect landmark data in camera frames.
    var poseLandmarker: PoseLandmarker?
    
    /// A buffer to hold the most recent image. Useful for if the `poseLandmarker` is still working on the previous image.
    var savedImage: MPImage?
    
    /// Flag for if the pose detector is currently working on proccessing an image into landmark data.
    var isWorking = false
    
    
    /// Regiesters the plugin with Flutter. Like an init function, creates a `PoseDetector` object for Flutter to use.
    /// and starts a thread for the object to detect images indefinitly.
    /// - Parameter registrar: Provides the plugin access to contextual information and register callbacks for various application events.
    public static func register(with registrar: FlutterPluginRegistrar)
    {
        let poseDetector = PoseDetector()
        
        let methodChannel = FlutterMethodChannel(name: "google_mediapipe_pose_detection", binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(poseDetector, channel: methodChannel)
        
        let eventChannel = FlutterEventChannel(name: "google_mediapipe_pose_detection_results", binaryMessenger: registrar.messenger())
        eventChannel.setStreamHandler(poseDetector)
        
        DispatchQueue.global().async
        {
            while(true)
            {
                poseDetector.detectSavedImage()
            }
        }
    }
    
    /// Handles incoming calls from Flutter.
    /// - Parameters:
    ///   - call: The method and data it wants to activate.
    ///   - result: A `FlutterResult` to send error messages back to Flutter.
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult)
    {
        switch call.method
        {
        case "startPoseDetector":
            handleDetection(call: call, result: result)
        case "closePoseDetector":
            result(nil)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    /// Takes the image data and converts it to the landmark data. Creates a PoseLandmarker if needed.
    /// - Parameters:
    ///   - call: The data Flutter sends about how the request should be interpeted.
    ///   - result: A `FlutterResult` to send error messages back to Flutter.
    public func handleDetection(call: FlutterMethodCall, result: @escaping FlutterResult)
    {
        guard let args = call.arguments as? [String: Any],
              let imageData = args["imageData"] as? [String: Any],
              let mpImage = InputImageConverter.visionImage(from: imageData, result: result),
              let options = args["options"] as? [String: Any],
              let mode = options["mode"] as? String
        else
        {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
            return
        }
        
        if poseLandmarker == nil
        {
            setupPoseLandmarker(options: options, result: result)
        }
        
        if poseLandmarker == nil
        {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Could not create poseLandmarker", details: nil))
            return
        }
        
        if mode == "image"
        {
            isWorking = true
            do
            {
                let detectionResult = try poseLandmarker!.detect(image: mpImage)
                
                isWorking = false
                
                let landmarkResult: [[[String: Any]]] = PoseDetector.convertPoseData(poseLandmarkerResult: detectionResult)
                
                result(landmarkResult)
            }
            catch
            {
                isWorking = false
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Could not detect image", details: nil))
            }
        }
        else if mode == "liveStream"
        {
            savedImage = mpImage
        }
        
    }
    
    /// Attempts to send a new image to the `poseLandmarker` for processing.
    public func detectSavedImage()
    {
        if !isWorking && savedImage != nil
        {
            isWorking = true
            do
            {
                let frameTime = Int(CACurrentMediaTime() * 1000)
                try poseLandmarker?.detectAsync(image:savedImage!, timestampInMilliseconds: frameTime)
                savedImage = nil
            }
            catch
            {
                savedImage = nil
                isWorking = false
            }
        }
    }
    
    /// Tries to create a new `PoseLandmarker` to use for detection.
    /// - Parameters:
    ///   - options: A dictionary defining how the `PoseLandmarker` should be created.
    ///   - result: A `FlutterResult` to send error messages back to Flutter.
    public func setupPoseLandmarker(options: [String: Any], result: @escaping FlutterResult)
    {
        guard let model = options["model"] as? String,
              let minPoseDetectionConfidence = options["minPoseDetectionConfidence"] as? Double,
              let minPosePresenceConfidence = options["minPosePresenceConfidence"] as? Double,
              let minPoseTrackingConfidence = options["minTrackingConfidence"] as? Double,
              let numPoses = options["numPoses"] as? Int,
              let mode = options["mode"] as? String
        else
        {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
            return
        }
        
        var modelPath = "assets/"
        
        switch model
        {
        case "full":
            modelPath += "pose_landmarker_full"
        case "lite":
            modelPath += "pose_landmarker_lite"
        case "heavy":
            modelPath += "pose_landmarker_heavy"
        default:
            modelPath += "pose_landmarker_full"
        }
        
        let key = FlutterDartProject.lookupKey(forAsset: modelPath, fromPackage: "google_mediapipe_pose_detection")
        let path: String? = Bundle.main.path(forResource: key, ofType: "task")
        
        if path == nil
        {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Cannot find task", details: nil))
            return
        }
        
        let options = PoseLandmarkerOptions()
        options.baseOptions.modelAssetPath = path!
        options.minPoseDetectionConfidence = Float(minPoseDetectionConfidence)
        options.minPosePresenceConfidence = Float(minPosePresenceConfidence)
        options.minTrackingConfidence = Float(minPoseTrackingConfidence)
        options.numPoses = numPoses
        
        if mode == "liveStream"
        {
            options.runningMode = .liveStream
            let processor = self
            options.poseLandmarkerLiveStreamDelegate = processor
        }
        else if mode == "image"
        {
            options.runningMode = .image
        }
        
        do
        {
            poseLandmarker = try PoseLandmarker(options: options)
        }
        catch
        {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Failed to create poseLandmarker", details: nil))
        }
    }
    
    /// Method for `FlutterStreamHandler` to link the stream to the object on startup.
    /// - Parameters:
    ///   - arguments: Unused.
    ///   - events: Provides a link from Flutter to Swift.
    /// - Returns: Always returns `nil` as no error can occur.
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError?
    {
        eventSink = events
        return nil
    }
    
    /// Method for `FlutterStreamHandler` to unlink the stream to the object on teardown.
    /// - Parameter arguments: Unused.
    /// - Returns: Always returns `nil` as no error can occur.
    public func onCancel(withArguments arguments: Any?) -> FlutterError?
    {
        eventSink = nil
        return nil
    }
    
    /// Converts the poses returned from the `PoseLandmarker` to a array of dictionarys of landmark variables to be sent to Flutter.
    /// - Parameter poseLandmarkerResult: The pose returned from `PoseLandmarker`.
    /// - Returns: A 3D structure that can be sent back to Flutter. First a list with each pose (useful if `numPoses` is more then `1`), then holds
    /// a map for each landmark in the pose. Finally, each landmark is a map of variables describing the landmark.
    public static func convertPoseData(poseLandmarkerResult: PoseLandmarkerResult) -> [[[String: Any]]]
    {
        var landmarkArray: [[Dictionary<String, Any>]] = []
        
        if !poseLandmarkerResult.landmarks.isEmpty
        {
            for pose in poseLandmarkerResult.landmarks
            {
                var landmarks: [Dictionary<String, Any>] = []
                
                for (index, landmark) in pose.enumerated()
                {
                    var landmarkMap: [String: Any] = [:]
                    
                    landmarkMap["type"] = index
                    landmarkMap["x"] = landmark.x
                    landmarkMap["y"] = landmark.y
                    landmarkMap["z"] = landmark.z
                    landmarkMap["presence"] = landmark.presence ?? 0.0
                    landmarkMap["visibility"] = landmark.visibility ?? 0.0
                    
                    landmarks.append(landmarkMap)
                }
                landmarkArray.append(landmarks)
            }
        }
        return landmarkArray
    }
}

/// Provides the class with the ability to interpret poses from the `PoseLandmarker`.
extension PoseDetector: PoseLandmarkerLiveStreamDelegate
{
    
    /// A method that the pose landmarker calls once it finishes performing
    /// landmarks detection in each input frame. Do not use directly.
    /// - Parameters:
    ///   - poseLandmarker: The object that detected the pose.
    ///   - result: the object representing the pose in frame.
    ///   - timestampInMilliseconds: A time value representing the detection.
    ///   - error: `nil` unless a detection failed.
    public func poseLandmarker(
        _ poseLandmarker: PoseLandmarker,
        didFinishDetection result: PoseLandmarkerResult?,
        timestampInMilliseconds: Int,
        error: Error?)
    {
        if error != nil
        {
            print(error?.localizedDescription ?? "Unkown Error")
            return
        }
        
        isWorking = false
        
        if result != nil
        {
            let landmarkList: [[[String: Any]]] = PoseDetector.convertPoseData(poseLandmarkerResult: result!)
            
            DispatchQueue.main.async
            {
                self.eventSink?(landmarkList)
            }
        }
    }
}
